### ISSUE TITLE
Possible false positive for OrderedDict.update with a dict

### ISSUE THREAD

**Issue by dimaqq** [microsoft/pyright#7186]
```py
from collections import OrderedDict

tmp = OrderedDict([("a", 1)])
tmp.update({"a": 42})
```

Works correctly on Py3.8..py3.12 (all I have on my system atm.)

Yet, `pyright` complains that:
```command
> pyright ord.py
./ord.py
  ./ord.py:4:1 - error: No overloads for "update" match the provided arguments (reportCallIssue)
  ./ord.py:4:13 - error: Argument of type "dict[str, int]" cannot be assigned to parameter "__m" of type "Iterable[tuple[Literal['a'], Literal[1]]]" in function "update"
    "Literal['a']" is incompatible with "tuple[Literal['a'], Literal[1]]" (reportArgumentType)
2 errors, 0 warnings, 0 informations
```

`pyright 1.1.349`
`Python 3.12.1`

**Comment by erictraut** [comment]
Pyright is working as designed here, so I don't consider this a bug. Refer to [this documentation](https://microsoft.github.io/pyright/#/type-inference?id=tuple-expressions) for details about how pyright infers the types of tuples.

**Comment by dimaqq** [comment]
🤔 well that's rather unexpected... let me expand a little on this:


```py
from collections import OrderedDict

a = "a"
one = 1
two = 2


def foo():
    tmp = OrderedDict({a: one})
    tmp.update({a: two})

    tmp2 = {a: one}
    tmp2.update({a: two})

    tmp3 = dict([(a, one)])
    tmp3.update({a: two})

    tmp4 = dict([[a, one]])
    tmp4.update({a: two})
```

No literals are used in constructors, instead local variables are used, thus the value of `a` and `one` could change, yet, constructing a dict (or ordered dict before) from a sequence seems to make pyright decide that the dict can only hold literals?


```command
> pyright ord.py
./ord.py
  ./ord.py:16:5 - error: No overloads for "update" match the provided arguments (reportCallIssue)
  ./ord.py:16:18 - error: Argument of type "dict[str, int]" cannot be assigned to parameter "__m" of type "Iterable[tuple[Literal['a'], Literal[1]]]" in function "update"
    "Literal['a']" is incompatible with "tuple[Literal['a'], Literal[1]]" (reportArgumentType)
  ./ord.py:18:12 - error: No overloads for "__init__" match the provided arguments (reportCallIssue)
  ./ord.py:18:17 - error: Argument of type "list[list[str | int]]" cannot be assigned to parameter "__iterable" of type "Iterable[list[bytes]]" in function "__init__"
    "Literal['a']" is incompatible with "bytes"
    "Literal[1]" is incompatible with "bytes" (reportArgumentType)
  ./ord.py:19:5 - error: No overloads for "update" match the provided arguments (reportCallIssue)
  ./ord.py:19:18 - error: Argument of type "dict[str, int]" cannot be assigned to parameter "__m" of type "Iterable[tuple[bytes, bytes]]" in function "update"
    "Literal['a']" is incompatible with "tuple[bytes, bytes]" (reportArgumentType)
6 errors, 0 warnings, 0 informations
```

- pyright thinks that `tmp3` cannot be updated?
  - creating a dict (or ordered dict) from a dict literal is OK here, dict can hold anything
  - creating a dict (or ordered dict) is somehow not OK, dict keys must be literals?
- pyright is confused about creating a dict from a list of lists (new)


If I add code that changes the module globals, the error for `tmp3` disappears:
```
foo()
a,one,two = 9,8,7
foo()
```
